pkg/wdk/services_aggregated_postebeef_result.go

```
@@ -0,0 +1,111 @@
+	AggregatedPostedTxIDServiceError AggregatedPostedTxIDStatus = "serviceError"
+)
+
+// AggregatedPostedTxID represents postBEEF result, aggregated from all broadcasters for particular TxID
+type AggregatedPostedTxID struct {
+	TxID              string
+	TxIDResults       []*PostedTxID
+	Status            AggregatedPostedTxIDStatus
+	SuccessCount      int
+	DoubleSpendCount  int
+	StatusErrorCount  int
+	ServiceErrorCount int
+	CompetingTxs      map[string]struct{}
+}
+
+// AggregatedPostBEEF is a map of AggregatedPostedTxID results, indexed by txid
+type AggregatedPostBEEF map[string]*AggregatedPostedTxID
+
+func (a AggregatedPostBEEF) getOrDefault(txid string) *AggregatedPostedTxID {
+	if agg, ok := a[txid]; ok {
+		return agg
+	}
+
+	agg := &AggregatedPostedTxID{
+		TxID:         txid,
+		CompetingTxs: make(map[string]struct{}),
+	}
+	a[txid] = agg
+
+	return agg
+}
+
+func (a AggregatedPostBEEF) summarize(txID string) {
+	agg, ok := a[txID]
+	if !ok {
+		agg = &AggregatedPostedTxID{
+			TxID:         txID,
+			CompetingTxs: make(map[string]struct{}),
+			Status:       AggregatedPostedTxIDServiceError,
+		}
+		a[txID] = agg
+		return
+	}
+
+	switch {
+	case agg.DoubleSpendCount > 0:
+		agg.Status = AggregatedPostedTxIDDoubleSpend
+	case agg.SuccessCount > 0:
+		agg.Status = AggregatedPostedTxIDSuccess
+	case agg.ServiceErrorCount > 0:
+		agg.Status = AggregatedPostedTxIDServiceError
+	default:
+		agg.Status = AggregatedPostedTxIDInvalidTx
+	}
+}
+
+func newAggregatedPostBEEF(results PostBeefResult, txids []string) AggregatedPostBEEF {
+	aggregatedTxs := make(AggregatedPostBEEF)
+
+	for _, result := range results {
+		if !result.Success() {
+			continue
+		}
+
+		mapped := make(map[string]*PostedTxID)
+		for _, txIDResult := range result.PostedBEEFResult.TxIDResults {
+			mapped[txIDResult.TxID] = &txIDResult
+		}
+
+		for _, txid := range txids {
+			txIDResult, ok := mapped[txid]
+			if !ok {
+				continue
+			}
```

hmmm... shouldn't we actually return an error, or panic - as it is unexpected behaviour I belive.

---

No it's not an error case. It could happen that some tx is not in the response. The `summarize` step, after checking all the results, will also create aggregated results for such transactions.